### PR TITLE
Improved foreign key detection

### PR DESCRIPTION
**Pull request by George3d6**

### CODE REVIEW

File: mindsdb_native/libs/helpers/text_helpers.py [hunk 1]
```diff
@@ -166,8 +167,8 @@ def is_foreign_key(data, column_name, data_subtype, other_potential_subtypes):
                 prev_val_length = len(str(val))
             elif len(str(val)) != prev_val_length:
                 is_same_length = False
-            else:
-                prev_val_length = len(str(val))
+
+            prev_val_length = len(str(val))
```
Comment (maximlopin) [inline mindsdb_native/libs/helpers/text_helpers.py:170]: The for loops can be simplified to this:
```
data = [str(x) for x in data]
all_same_length = all(len(data[0]) == len(x) for x in data)
uuid_charset = set('0123456789abcdef-')
is_uuid = all(set(x).issubset(uuid_charset) for x in data)
data_looks_like_id = (all_same_length and is_uuid)
```

**Review by StpMax — CHANGES_REQUESTED** [review]
I made mistake in original issue, should be: if in dataset exists serial column, then in most cases trained predictor will be bad.
I checked branch, it fix issue for column with name 'id', but it still possible get bad predictor if serial column exists.

**Comment by George3d6** [comment]
@StpMax what is the serial column in pandas you're have issues with identified as when stringified ? Is it `'serial'` or some other name ?

File: mindsdb_native/libs/helpers/text_helpers.py (hunk 1 above)
Comment (George3d6) [reply mindsdb_native/libs/helpers/text_helpers.py:170]: made some changes based on this.

File: mindsdb_native/libs/helpers/text_helpers.py [hunk 2]
```diff
@@ -158,10 +158,8 @@ def is_foreign_key(data, column_name, data_subtype, other_potential_subtypes):
             is_uuid = True
             is_same_length = True
 
-            for char in str(val):
-                if char not in ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
-                                'a', 'b', 'c', 'd', 'e', 'f', '-']:
-                    is_uuid = False
+            uuid_charset = set('0123456789abcdef-')
+            set(val).issubset(uuid_charset)
```
Comment (StpMax) [inline mindsdb_native/libs/helpers/text_helpers.py]: when i train 'home_rentals' i get error in this row:
`ERROR:mindsdb-logger-1cc370aa-dbab-11ea-84a9-2c56dc4ecd27:/home/maxs/dev/mdb/venv_new/sources/mindsdb_native/mindsdb_native/libs/controllers/transaction.py:178 - 'float' object is not iterable`
it try execute `set(1906.0)` and fail

**Review by StpMax — CHANGES_REQUESTED** [review]

**Comment by StpMax** [comment]
> @StpMax what is the serial column in pandas you're have issues with identified as when stringified ? Is it `'serial'` or some other name ?

Under 'serial' i mean database serial type (int autoincrement in most cases), with any name of column. I made test file to add clarity:
```
import pandas as pd
import mindsdb as mdb
from sklearn.utils import shuffle

ROWS = 5000

df = pd.read_csv('home_rentals.csv', sep=',')
df = df[:ROWS]

p1 = mdb.Predictor(name='dt_1')
p1.learn(from_data=df, to_predict='rental_price', use_gpu=False)

df['index'] = 0
for i in range(ROWS):
    df.loc[i, 'index'] = i
p2 = mdb.Predictor(name='dt_2')
p2.learn(from_data=df, to_predict='rental_price', use_gpu=False)

df_random_sort = shuffle(df)
p3 = mdb.Predictor(name='dt_3')
p3.learn(from_data=df_random_sort, to_predict='rental_price', use_gpu=False)

df = df.astype({'number_of_rooms': 'string'})
p4 = mdb.Predictor(name='dt_4')
p4.learn(from_data=df, to_predict='rental_price', use_gpu=False)

df_random_sort = df_random_sort.astype({'number_of_rooms': 'string'})
p5 = mdb.Predictor(name='dt_5')
p5.learn(from_data=df_random_sort, to_predict='rental_price', use_gpu=False)

when_data = {'sqft': 1000, 'location': 'good'}
r1 = p1.predict(when_data=when_data, run_confidence_variation_analysis=True, use_gpu=False)
r2 = p2.predict(when_data=when_data, run_confidence_variation_analysis=True, use_gpu=False)
r3 = p3.predict(when_data=when_data, run_confidence_variation_analysis=True, use_gpu=False)
r4 = p4.predict(when_data=when_data, run_confidence_variation_analysis=True, use_gpu=False)
r5 = p5.predict(when_data=when_data, run_confidence_variation_analysis=True, use_gpu=False)

print("\nDF without 'index' column:")
print(r1._data)
print(r1._extra_insights)
print("\nDF with 'index' column dtype=int64, sorted:")
print(r2._data)
print(r2._extra_insights)
print("\nDF with 'index' column dtype=int64, not sorted:")
print(r3._data)
print(r3._extra_insights)
print("\nDF with 'index' column dtype=string, sorted:")
print(r4._data)
print(r4._extra_insights)
print("\nDF with 'index' column dtype=string, not sorted:")
print(r5._data)
print(r5._extra_insights)
```
i run it several times with different ROWS, the output is:

ROWS = 5000:
```
DF without 'index' column:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2402.819613540836], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.995], 'rental_price_confidence_range': [[4277.541001393053, 119.4065925383611]], 'rental_price_confidence': [0.9365564773656152], 'model_rental_price': [2402.819613540836]}
{'rental_price': {'if_missing': [{'sqft': 2369.339341832317}, {'location': 1762.4974556957745}]}}

DF with 'index' column dtype=int64, sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [3243.906970310856], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.98], 'rental_price_confidence_range': [[-1075.6573042067605, 3130.5893922982646]], 'rental_price_confidence': [0.9982137662428189], 'model_rental_price': [3243.906970310856]}
{'rental_price': {'if_missing': [{'sqft': -7177.489720342722}, {'location': 2504.182094516758}]}}

DF with 'index' column dtype=int64, not sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2490.9147672495174], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.9889887667511474], 'rental_price_confidence_range': [[-189.9550016637428, 1655.2019216921817]], 'rental_price_confidence': [0.00031977461519471114], 'model_rental_price': [2490.9147672495174]}
{'rental_price': {'if_missing': [{'sqft': -7038.974125529655}, {'location': 2093.7148105915426}]}}

DF with 'index' column dtype=string, sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [3243.906970310856], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.98], 'rental_price_confidence_range': [[-1075.6573042067605, 3130.5893922982646]], 'rental_price_confidence': [0.9982137662428189], 'model_rental_price': [3243.906970310856]}
{'rental_price': {'if_missing': [{'sqft': -7177.489720342722}, {'location': 2504.182094516758}]}}

DF with 'index' column dtype=string, not sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2490.9147672495174], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.9889887667511474], 'rental_price_confidence_range': [[-189.9550016637428, 1655.2019216921817]], 'rental_price_confidence': [0.00031977461519471114], 'model_rental_price': [2490.9147672495174]}
{'rental_price': {'if_missing': [{'sqft': -7038.974125529655}, {'location': 2093.7148105915426}]}}
```

ROWS = 3000
```
DF without 'index' column:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2383.988019941377], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.98], 'rental_price_confidence_range': [[2230.0710153875116, 2877.2533782434434]], 'rental_price_confidence': [0.513802150285053], 'model_rental_price': [2383.988019941377]}
{'rental_price': {'if_missing': [{'sqft': 2119.479866347027}, {'location': 2171.7866958230356}]}}

DF with 'index' column dtype=int64, sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2686.964349252861], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.95], 'rental_price_confidence_range': [[484.39339204401085, -811.6994827756253]], 'rental_price_confidence': [0.9858713705596995], 'model_rental_price': [2686.964349252861]}
{'rental_price': {'if_missing': [{'sqft': -5935.540481848001}, {'location': 2179.5797030544754}]}}

DF with 'index' column dtype=int64, not sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2607.3730153650613], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.9], 'rental_price_confidence_range': [[1084.6966451450394, -895.8120437691678]], 'rental_price_confidence': [0.9108324436497988], 'model_rental_price': [2607.3730153650613]}
{'rental_price': {'if_missing': [{'sqft': -7522.648316485373}, {'location': 2363.451513195079}]}}

DF with 'index' column dtype=string, sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2686.964349252861], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.95], 'rental_price_confidence_range': [[484.39339204401085, -811.6994827756253]], 'rental_price_confidence': [0.9858713705596995], 'model_rental_price': [2686.964349252861]}
{'rental_price': {'if_missing': [{'sqft': -5935.540481848001}, {'location': 2179.5797030544754}]}}

DF with 'index' column dtype=string, not sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [2607.3730153650613], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.9], 'rental_price_confidence_range': [[1084.6966451450394, -895.8120437691678]], 'rental_price_confidence': [0.9108324436497988], 'model_rental_price': [2607.3730153650613]}
{'rental_price': {'if_missing': [{'sqft': -7522.648316485373}, {'location': 2363.451513195079}]}}
```

ROWS = 1000
```
DF without 'index' column:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [3372.058768219188], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.995], 'rental_price_confidence_range': [[1716.9742604710357, -2641.7502138707787]], 'rental_price_confidence': [0.103247904461855], 'model_rental_price': [3372.058768219188]}
{'rental_price': {'if_missing': [{'sqft': -830.7834747841516}, {'location': 3777.6846484536754}]}}

DF with 'index' column dtype=int64, sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [-1786.5299656460588], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.995], 'rental_price_confidence_range': [[-1361.4347155155997, 424.65440447559376]], 'rental_price_confidence': [0.00533167290728434], 'model_rental_price': [-1786.5299656460588]}
{'rental_price': {'if_missing': [{'sqft': 927.7621133019637}, {'location': -3037.734925786088}]}}

DF with 'index' column dtype=int64, not sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [-2175.6183433021824], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.98], 'rental_price_confidence_range': [[3324.850533926246, -2589.306111670491]], 'rental_price_confidence': [0.9422577515317097], 'model_rental_price': [-2175.6183433021824]}
{'rental_price': {'if_missing': [{'sqft': 2365.142726804085}, {'location': -3497.0359417219947}]}}

DF with 'index' column dtype=string, sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [-1786.5299656460588], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.995], 'rental_price_confidence_range': [[-1361.4347155155997, 424.65440447559376]], 'rental_price_confidence': [0.00533167290728434], 'model_rental_price': [-1786.5299656460588]}
{'rental_price': {'if_missing': [{'sqft': 927.7621133019637}, {'location': -3037.734925786088}]}}

DF with 'index' column dtype=string, not sorted:
{'number_of_rooms': ['None'], 'number_of_bathrooms': ['None'], 'sqft': [1000], 'location': ['good'], 'days_on_market': [None], 'initial_price': [None], 'neighborhood': ['None'], 'rental_price': [-2175.6183433021824], 'index': [None], '__observed_rental_price': [None], 'rental_price_model_confidence': [0.98], 'rental_price_confidence_range': [[3324.850533926246, -2589.306111670491]], 'rental_price_confidence': [0.9422577515317097], 'model_rental_price': [-2175.6183433021824]}
{'rental_price': {'if_missing': [{'sqft': 2365.142726804085}, {'location': -3497.0359417219947}]}}
```

I used lightwood=0.34.0 and this branch of mindsb_native when run file.
What problems i see in output:
1. if any 'serial' column exists (with int or string type, no matter), then predicted rental_price value (or values from 'if_missing') bad (this is original issue).
2. `rental_price_confidence` for datasets with 'serial' column in some cases very low.
3. rental_price_confidence_range everywhere bad.

**Comment by George3d6** [comment]
@StpMax I added a check for serial, let me know if it works, not much else we can easily do right now I'm afraid :/

File: mindsdb_native/libs/helpers/text_helpers.py (hunk 2 above)
Comment (maximlopin) [reply mindsdb_native/libs/helpers/text_helpers.py]: it should be `set(str(val)).issubset(uuid_charset)`

**Pull request merged**